### PR TITLE
Add intermittent Timestamping to Votes

### PR DESCRIPTION
**Pull request by CriesofCarrots**
#### Problem
Users have no way to know when a Solana block was produced in real-world time.
We have a design proposal to fix that issue... here is the first step of the implementation!

#### Summary of Changes
- Add VoteState::last_timestamp field
- Add `Option<timestamp>` field to Vote
- Trigger intermittent timestamp in replay_stage::handle_votable_bank. Currently set to approximately every 30min (4500 slots, with current defaults). First timestamp occurs on slot 1
- Stash last_timestamp in Tower to keep validator timestamping regularly despite potentially missing slots

Toward #7115 

@mvines , I believe we wanted this in 0.21 too, correct? If so, I'll also cherry-pick the design proposal for completeness.

### CODE REVIEW

File: programs/vote/src/vote_state.rs [hunk 1]
```diff
@@ -423,6 +452,17 @@ pub fn process_vote(
...
     vote_state.process_vote(vote, slot_hashes, clock.epoch)?;
+    if let Some(timestamp) = vote.timestamp {
+        let slot = vote
+            .slots
+            .iter()
+            .max()
+            .expect("vote must contain at least one slot");
+        let res = vote_state.process_timestamp(*slot, timestamp);
+        if let Err(err) = res {
+            warn!("Timestamp: {:?}, Error: {:?}", (slot, timestamp), err);
```
Comment (CriesofCarrots) [inline programs/vote/src/vote_state.rs]: I assume we would not want to fail the whole Vote on an invalid timestamp. Correct?
Maybe this should be even quieter (info?)

File: programs/vote/src/vote_state.rs [hunk 2]
```diff
@@ -423,6 +452,17 @@ pub fn process_vote(
     verify_authorized_signer(&vote_state.authorized_voter, signers)?;
 
     vote_state.process_vote(vote, slot_hashes, clock.epoch)?;
+    if let Some(timestamp) = vote.timestamp {
+        let slot = vote
+            .slots
+            .iter()
+            .max()
+            .expect("vote must contain at least one slot");
```
Comment (rob-solana) [inline programs/vote/src/vote_state.rs]: you mustn't expect() in these programs, you must return an error....

File: programs/vote/src/vote_state.rs (hunk 1 above)
Comment (rob-solana) [reply programs/vote/src/vote_state.rs]: same level of logging as other rejected votes, IMHO

I think I'd want the timestamp to be valid to let in a vote, yes.

File: sdk/src/timing.rs [hunk 3]
```diff
@@ -47,6 +47,12 @@ pub fn slot_duration_from_slots_per_year(slots_per_year: f64) -> Duration {
     Duration::from_nanos(slot_in_ns as u64)
 }
 
+/// From slots per year to slots per some minutes, rounded
+pub fn slots_per_interval(slots_per_year: f64, interval_mins: u64) -> u64 {
```
Comment (rob-solana) [inline sdk/src/timing.rs]: this would be minutes_as_slots() ?  why "interval"?

Comment (rob-solana) [reply sdk/src/timing.rs]: also: above, you mix ticks and slots line 41 "/// From slots per year to tick_duration", but the function is called `slot_duration_`

File: programs/vote/src/vote_state.rs [hunk 4]
```diff
@@ -334,6 +343,26 @@ impl VoteState {
             }
         }
     }
+
+    fn check_timestamp(&self, slot: Slot, timestamp: UnixTimestamp) -> Result<(), VoteError> {
```
Comment (rob-solana) [inline programs/vote/src/vote_state.rs]: can you just fold this into `process_timestamp()` ?

File: core/src/replay_stage.rs [hunk 5]
```diff
@@ -652,8 +653,14 @@ impl ReplayStage {
             let node_keypair = cluster_info.read().unwrap().keypair.clone();
 
             // Send our last few votes along with the new one
+            let mut last_vote = tower.last_vote();
+            let timestamp_interval =
+                slots_per_interval(bank.slots_per_year(), DEFAULT_TIMESTAMP_INTERVAL_MINS);
+            if bank.slot() % timestamp_interval == 1 {
```
Comment (rob-solana) [inline core/src/replay_stage.rs]: when is  `slot % slots_per_30_min == 1` ?

File: core/src/replay_stage.rs [hunk 6]
```diff
@@ -652,8 +653,14 @@ impl ReplayStage {
             let node_keypair = cluster_info.read().unwrap().keypair.clone();
 
             // Send our last few votes along with the new one
+            let mut last_vote = tower.last_vote();
+            let timestamp_interval =
```
Comment (rob-solana) [inline core/src/replay_stage.rs]: seems like we could just pick some hardcoded slot frequency instead of doing this calculation over and over again...

File: sdk/src/clock.rs [hunk 7]
```diff
@@ -35,6 +35,8 @@ pub const NUM_CONSECUTIVE_LEADER_SLOTS: u64 = 4;
 
 pub const DEFAULT_MS_PER_SLOT: u64 = 1_000 * DEFAULT_TICKS_PER_SLOT / DEFAULT_TICKS_PER_SECOND;
 
+pub const DEFAULT_TIMESTAMP_INTERVAL_MINS: u64 = 30;
```
Comment (rob-solana) [inline sdk/src/clock.rs]: this belongs in the vote program, I think

File: core/src/replay_stage.rs (hunk 5 above)
Comment (rob-solana) [reply core/src/replay_stage.rs]: slot_per_30_min is a huge number (30*60/0.4 =4500)...  we don't vote on every bank or consider voting on every bank, so we could go a **_long_** time between updates.

Probably need to keep the last timestamp in Tower and just do a Duration:: check for 5 minutes or something...

File: programs/vote/src/vote_state.rs (hunk 2 above)
Comment (CriesofCarrots) [reply programs/vote/src/vote_state.rs]: Duh, forgot where I was. Thanks!

File: programs/vote/src/vote_state.rs (hunk 4 above)
Comment (CriesofCarrots) [reply programs/vote/src/vote_state.rs]: Yep. I thought it might help with unit testing, but it is unnecessarily verbose.

File: core/src/replay_stage.rs (hunk 5 above)
Comment (CriesofCarrots) [reply core/src/replay_stage.rs]: > when is `slot % slots_per_30_min == 1` ?

With current slots_per_year, that happens on slot 1, slot 4501, slot 9001, etc...
I chose 1 instead of 0 so that we'd get a timestamp right away on cluster boot.

The reason I went with a specific slot instead of a duration is that it makes it easier to find these votes in the ledger, which we need to do to calculate a particular block time in history. I like your idea of keeping the last timestamp in Tower, however... Do you have any thoughts on how to find the most recent timestamp Votes from the ledger without scanning the entire duration's-worth of slots?

Regarding the specific interval, 30min was proposed in our initial discussion to keep the extra bytes on the wire small. We can certainly drop the interval, if we are okay

File: sdk/src/timing.rs (hunk 3 above)
Comment (CriesofCarrots) [reply sdk/src/timing.rs]: > this would be minutes_as_slots() ? why "interval"?

No good reason. minutes_as_slots() is much better!

File: core/src/replay_stage.rs (hunk 5 above)
Comment (rob-solana) [reply core/src/replay_stage.rs]: on wire size: is it really smaller for us to serialize a struct with an Option<timestamp> than a struct with a timestamp?

the thing is, this code doesn't necessarily get to see bank 1 or bank 4501...

I don't have a good idea for how to troll the ledger offhand.  also: the failed votes with bogus timestamps will still be in the ledger...

Comment (CriesofCarrots) [reply core/src/replay_stage.rs]: > on wire size: is it really smaller for us to serialize a struct with an Option than a struct with a timestamp?

None case: 1 byte vs 8 bytes?

> the thing is, this code doesn't necessarily get to see bank 1 or bank 4501...

Yeah, I get that.

> I don't have a good idea for how to troll the ledger offhand. also: the failed votes with bogus timestamps will still be in the ledger...
Yep, but we can check statuses.
I'll think on efficient trolling...

Comment (rob-solana) [reply core/src/replay_stage.rs]: > > on wire size: is it really smaller for us to serialize a struct with an Option than a struct with a timestamp?
> 
> None case: 1 byte vs 8 bytes?
>

bincode::serialize() is sneaky!
 
> > the thing is, this code doesn't necessarily get to see bank 1 or bank 4501...
> 
> Yeah, I get that.

so the thinking is that *somebody* is gonna hit the 1s?


> > I don't have a good idea for how to troll the ledger offhand. also: the failed votes with bogus timestamps will still be in the ledger...
> 
> Yep, but we can check statuses.
> I'll think on efficient trolling...

if Tower remembered the last vote that had a timestamp, you could do:

```
if last_timestamp_slot % interval < bank.slot() % interval {
}
```
then everybody'd be sending a timestamp every half hour, offset by several slots

Comment (CriesofCarrots) [reply core/src/replay_stage.rs]: > so the thinking is that _somebody_ is gonna hit the 1s?

No, you're right that this approach needs to change.

> if Tower remembered the last vote that had a timestamp, you could do:
> 
> ```
> if Tower.last_timestamp_slot % interval < bank.slot() % interval {
> }
> ```
> 
> then everybody'd be sending a timestamp every half hour, offset by several slots

Yes, nice. Just landing on that `<` myself.

**Comment by CriesofCarrots** [comment]
@rob-solana , thanks for the review! How does this look now?

File: core/src/replay_stage.rs [hunk 8]
```diff
@@ -652,8 +653,14 @@ impl ReplayStage {
             let node_keypair = cluster_info.read().unwrap().keypair.clone();
 
             // Send our last few votes along with the new one
+            let mut last_vote = tower.last_vote();
+            if tower.needs_timestamp(bank.slot()) {
+                let timestamp = Utc::now().timestamp();
+                last_vote.timestamp = Some(timestamp);
+                tower.record_recent_timestamp(bank.slot(), timestamp);
```
Comment (rob-solana) [inline core/src/replay_stage.rs]: nice, you could combine the needs() and record() into a single function that returns an Option that you'd stuff directly into last_vote...  you could also have last_vote() take a slot and do all the work...

Comment (CriesofCarrots) [reply core/src/replay_stage.rs]: Actually, if I do it that way, I don't even need the Slot, as I can take it from the last_vote. Rad.
I'm going to add a separate method, though, since `last_vote()` is used to print out the last vote a few places, and I don't want it updating the timestamp in those cases.

File: programs/vote/src/vote_state.rs [hunk 9]
```diff
@@ -334,6 +346,21 @@ impl VoteState {
             }
         }
     }
+
+    pub fn process_timestamp(
+        &mut self,
+        slot: Slot,
+        timestamp: UnixTimestamp,
+    ) -> Result<(), VoteError> {
+        if (slot < self.last_timestamp.0 || timestamp < self.last_timestamp.1)
```
Comment (mvines) [inline programs/vote/src/vote_state.rs]: I'm non a fan of using tuples for public fields, I'd much prefer a struct because:

1. I keep having to refer back to the struct because I forget what .0 and .1 are
2. I get confused when I see things like `slot == self.last_timestamp.0`.  My brain likes to ignore the insignificant looking `.0` suffix and then wonders why we're comparing a slot to a timestamp.

File: ci/localnet-sanity.sh [hunk 10]
```diff
@@ -365,7 +365,7 @@ while [[ $iteration -le $iterations ]]; do
   echo "--- Wallet sanity ($iteration)"
   (
     set -x
-    timeout 90s scripts/wallet-sanity.sh --url http://127.0.0.1"$walletRpcPort"
+    timeout 105s scripts/wallet-sanity.sh --url http://127.0.0.1"$walletRpcPort"
```
Comment (mvines) [inline ci/localnet-sanity.sh]: Does this belong in a separate PR?

File: programs/vote/src/vote_state.rs [hunk 11]
```diff
@@ -25,17 +25,26 @@ pub const INITIAL_LOCKOUT: usize = 2;
 //  smaller numbers makes
 pub const MAX_EPOCH_CREDITS_HISTORY: usize = 64;
 
+// Frequency of timestamp Votes
+pub const DEFAULT_TIMESTAMP_SLOTS: u64 = 20;
```
Comment (mvines) [inline programs/vote/src/vote_state.rs]: `TIMESTAMP_SLOT_INTERVAL` perhaps?   No need for DEFAULT_ as there's no other possible value at the moment, and this the interval between timestamps as measured in slots

Comment (mvines) [reply programs/vote/src/vote_state.rs]: Oh can you also please add some wording around why 20 was chosen here?

**Comment by mvines** [comment]
lgtm, just nits.   Also we don't need this on v0.21 so I removed the label

**Comment by CriesofCarrots** [comment]
@mvines , does this look okay to land?
